lib/model/message.dart

```diff
@@ -180,6 +183,64 @@ class MessageStoreImpl extends PerAccountStoreBase with MessageStore, _OutboxMes
     _disposed = true;
   }
 
+  static const _markReadOnScrollBatchSize = 1000;
+  static const _markReadOnScrollDebounceDuration = Duration(milliseconds: 500);
+  final _markReadOnScrollQueue = _MarkReadOnScrollQueue();
+  bool _markReadOnScrollBusy = false;
+
+  /// Returns true on success, false on failure.
+  Future<bool> _sendMarkReadOnScrollRequest(Iterable<int> toSend) async {
+    assert(toSend.isNotEmpty);
+
+    // TODO(#1581) mark as read locally for latency compensation
+    //   (in Unreads and on the message objects)
+    try {
+      await updateMessageFlags(connection,
+        messages: toSend.toList(),
+        op: UpdateMessageFlagsOp.add,
+        flag: MessageFlag.read);
+    } on ApiRequestException {
+      // TODO(#1581) un-mark as read locally?
+      return false;
+    }
+    return true;
+  }
+
+  @override
+  void markReadOnScroll(Iterable<int> messageIds) async {
+    assert(!_disposed);
+    final anyAdded = _markReadOnScrollQueue.add(messageIds);
+    if (!anyAdded) return;
+    if (_markReadOnScrollBusy) return;
```

Comment:
Is the anyAdded condition needed? If the queue is nonempty but the "busy" flag is false, then nothing's working on the queue; so then it seems good to start doing so, even if this call didn't happen to add anything new to the queue.

(Also it seems like a bug if that situation happens in the first place.)

Reply:
Without that condition, the `add` implementation can also get a bit simpler.